File: Solutions/948_Bag_of_Tokens/948_Bag_of_Tokens.js
var bagOfTokensScore = function(tokens, P) {
    let score = 0;
    tokens = tokens.sort((a, b) => a - b);

    while (tokens.length) {

        if (tokens[0] > P) {
            return score;
        }
        while (tokens.length && tokens[0] <= P) {
            P -= tokens.shift();
            score += 1
        }

        if (tokens.length > 1 && score) {
            score -= 1;
            P += tokens.pop();
        }
    }

    return score;
};

//Third fastest
/**
 * @param {number[]} tokens
 * @param {number} P
 * @return {number}
 */
var bagOfTokensScore = function(tokens, P) {
    let score =0;
    tokens.sort((x,y)=>x-y);
    
    
    for(let l=0,r=tokens.length-1;l<=r;){
        //if(P<tokens[l] && score===0 ) break;//优化 当能量低于剩下的最小令牌且score=0时 提前退出循环
        
        if(P>=tokens[l]){//能量>=剩下的最小能量 用能量换分
            score++;
            P-=tokens[l++];
        }else if(score>=1 && tokens[r] > tokens[l]){//能量不足以换分 看看能否用分换能量
            score--;
            P+=tokens[r--]
        }else{//能量不足以换分 又无法用分换能量
            // l++
            break;
        }
    }
    
    return score;
};

//Second fastest
var bagOfTokensScore = function(tokens, P) {
    tokens.sort((a,b) => a - b); // order by power
    const score = findMaxScore(tokens, P, 0);
    return score;
  };
  
  function findMaxScore(tokens, P, score) {
    if (tokens.length === 0) return score;
  
    // three actions
    // buy
    // flip
    // do nothing
    let lessTokens = tokens;
  
    if (P >= tokens[0]) {
      let affordableIndex = 0;
      let newP = P;
      do {
        lessTokens = tokens.slice(affordableIndex + 1);
        newP = newP - tokens[affordableIndex];
        affordableIndex += 1;
      } while (newP >= tokens[affordableIndex] && affordableIndex < tokens.length);
      return Math.max(score, findMaxScore(lessTokens, newP, score + affordableIndex));
    } else if (score >= 1) {
      let removedIndex = tokens.length - 1;
      let newP = P;
      let newScore = score;
      do {
        lessTokens = tokens.slice(0, removedIndex);
        newP = newP + tokens[removedIndex];
        removedIndex -= 1;
        newScore -= 1;
      } while(newP < tokens[0] && removedIndex > 0 && newScore >= 1);
      return Math.max(score, findMaxScore(lessTokens, newP, newScore));
    }
    return score;
  }

//Fastest
/**
 * @param {number[]} tokens
 * @param {number} P
 * @return {number}
 */
var bagOfTokensScore = function(tokens, P) {
    let score = 0;
    tokens = tokens.sort((a,b) => a-b);
    
    while (tokens.length) {
        
      if (tokens[0] > P) {
        return score;
      }
        while (tokens.length && tokens[0] <= P) {
          P -= tokens.shift();
          score += 1
        }
      
        if (tokens.length > 1 && score) {
          score -= 1;
          P += tokens.pop();
        }
    }
    
    return score;
  };

File: Solutions/948_Bag_of_Tokens/948_Bag_of_Tokens.cpp
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int P) {
        
        if(tokens.size() == 0)   /// Base case
            return 0;
        
        sort(tokens.begin(),tokens.end());
        if(P < tokens[0])   /// if P < minimum token value, we can't collect any points.
            return 0;
        int n = tokens.size();
        vector<int> prefix_sum(n,0);
        
        prefix_sum[0] = tokens[0];
        for(int i = 1;i<n;i++)
            prefix_sum[i] = prefix_sum[i-1] + tokens[i];
        
        int ans = lower_bound(prefix_sum.begin(),prefix_sum.end(),P) - prefix_sum.begin();   /// Initialize answer with the value of points we can collect without playing any face down card.
        
        int power = P;
        for(int i = n-1 ; i >= n/2; i--){     /// Running the loop till n/2 because face-up cards should be greater than face down to collexct points
             power += tokens[i];
          
            int points = lower_bound(prefix_sum.begin(),prefix_sum.end(),power) - prefix_sum.begin() - (n-i);    /// Finding the number of points for a particular value of power.
            if(binary_search(prefix_sum.begin(),prefix_sum.end(),power)){
                points = points+1;    /// Handling the case when Power of both the subsets are equal.
            }
           
            ans = max(ans,points);
        }
        return ans;
         
    }
};

//Third fastest
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int P) {
        if(tokens.empty()) return 0;
        std::sort(tokens.begin(), tokens.end());
        int p1 = 0, p2 = int(tokens.size()) - 1;
        int score = 0;
        int maxScore = 0;
        while(p1 <= p2)
        {
            if(tokens[p1] <= P)
            {
                P -= tokens[p1++];
                score++;
                maxScore = std::max(maxScore, score);
            }
            else
            {
                if(score == 0)
                {
                    break;
                }
                else
                {
                    score--;
                    P += tokens[p2--];
                }
            }
        }
        return maxScore;
    }
};

//Second fastest
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int P) {
        sort(tokens.begin(), tokens.end());
        int res = 0, points = 0, i = 0, j = tokens.size() - 1;
        while (i <= j) {
            if (P >= tokens[i]) {
                P -= tokens[i++];
                res = max(res, ++points);
            } else if (points > 0) {
                points--;
                P += tokens[j--];
            } else {
                break;
            }
        }
        return res;
    }
};

//Fastest
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int P) {
        sort(tokens.begin(), tokens.end());
        if (tokens.size() == 0 || P < tokens[0]) return 0;
        int curPoints = 0;
        int i = 0;
        int j = tokens.size()-1;
        
        while(i<=j)
        {
            if(P >= tokens[i])
            {
                curPoints++;
                P -= tokens[i];
                //maxPoints = max(maxPoints, curPoints);
                i++;
            }
            else{ if( j - i > 0)
            {
                P += tokens[j];
                curPoints--;
                j--;
            }
            else
                break;
                }
        }
        
        return curPoints;
    }
};